Adding a page (with demos) of Hero animations.

Sfshaza opened the pull request:
Please take a gander, @HansMuller!

STAGED: https://sz-flutter.firebaseapp.com/animations/shared-element-transitions/

By the way, Travis will definitely fail on a couple counts:
- MaterialRectCenterArcTween is not yet in the alpha release, so the analyzer will fail on the examples that use it.
- I've added some examples, and moved some examples, so links to those examples will fail. (They will be correct AFTER the PR lands.)
- In a recent PR, Adam modified the code (FractionalOffset => Alignment) in a couple examples. This PR will cause collisions - I can sort those out later.

Sfshaza commented:
STAGED: https://sz-flutter.firebaseapp.com/animations/shared-element-transitions/

Sfshaza commented on _includes/code/animation/flying_transition/pubspec.yaml:
```diff
@@ -1,4 +1,4 @@
-name: shared_element_transition
+name: flying_shared_element_transition
```
Suggestions? At one point I called this example shared_element_transition, but both this and the radial examples are shared_element_transitions.

HansMuller replied:
Maybe you could call this demo "heroic_flight" since (if I'm right about which demo you're referring to) it's just the basic hero flight. It may be asking too much to try and come up with names that can be read like descriptive expressions.

If that made sense then perhaps the other demo would be expanding_heroic_flight.

Sfshaza replied:
I don't love how vague "hero" really is. To many people, a "hero" is a banner image on a web page.

Sfshaza commented on animations/index.md:
```diff
@@ -105,13 +105,15 @@ In this pattern, the user selects an element&mdash;often an
-demo includes an example of a hero animation.
+demo includes an example of a Hero animation.
 {% comment %}
 According to Hans, the demo in the Flutter Gallery that is called
```
Yeah, we don't need that hidden comment.  ;)

Sfshaza commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,649 @@
+route to another. This guide describes the first example.<br><br>
+
+[flying_transition](https://github.com/flutter/website/tree/master/_includes/code/animation/flying_transition)
+: A movement transition as described in the Material motion spec.
```
Except that this is describing a standard Hero transition, not a radial transformation.

Sfshaza commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,649 @@
+The following diagram shows the clipped image at the beginning
+(`t = 0.0`), and the end (`t = 1.0`) of the animation.
+
+<img src="images/radial-transition.png" alt="visual diagram of
```
I'm going to go over with you about this tomorrow, when I'm in MTV.

Sfshaza commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,649 @@
+shapes intersect. At the beginning of the transition,
+the shapes intersect as a circle
+([ClipOval](https://docs.flutter.io/flutter/widgets/ClipOval-class.html)).
+At the end, the image and circle clip have scaled up, and the
```
It sure LOOKS bigger!!!

HansMuller replied:
Sorry. I read the sentence as the "the image clip and the circle clip have scaled up".

I'd say: "At the end of the transition the intersection of the circular and rectangular clips yield a rectangle that's the same size as the Hero widget. In other words, at the end of the transition the image is no longer clipped.

InMatrix commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,649 @@
+(or using the device's back-to-previous-screen gesture)
+returns the user to the original screen.
+
+<center><iframe width="560" height="315" src="https://www.youtube.com/embed/uF7_JfhBsww" frameborder="0" allowfullscreen></iframe></center>
```
Hope we can find a better solution than YouTube. It's really not ideal. In addition to showing a wide player for a vertical video, it suggests irrelevant videos after the demo video finishes. It can be very distracting. See an example below:

![image](https://user-images.githubusercontent.com/348942/31424168-a4b378c6-ae0d-11e7-8fe9-2ceecb91e235.png)

Sfshaza replied:
I spoke to Seth about this last week, @InMatrix. He said that if we can find a place to host the content, we can present a more customized solution (in portrait mode, for example), but it's more work to set up. For now, I'm taking the easy route but, I agree, we need a better solution.

InMatrix commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,649 @@
+
+[Create a new Flutter example](/getting-started/) and
+update it using the files from the
+[GitHub directory.](https://github.com/flutter/website/tree/master/_includes/code/animation/flying_transition)
```
The flying_transition folder didn't exist when I checked.

Sfshaza commented:
Yes, it won't until this PR lands. I've made lots of edits as per Hans, and
will talk to him today about a couple of things before I update this PR.
Thanks!

On Tue, Oct 10, 2017 at 11:17 PM, Tao Dong <notifications@github.com> wrote:

> *@InMatrix* commented on this pull request.
> ------------------------------
>
> In animations/shared-element-transitions/index.md
> <https://github.com/flutter/website/pull/693#discussion_r143920151>:
>
> > +  described in this guide.
> +
> +</aside>
> +
> +Flying an element from one route to another, one of the most
> +common types of shared element transitions, is easy to implement
> +using Flutter's Hero widget. When using MaterialPageRoute
> +to specify the new route, the element flies along a curved path,
> +as described by the [Material Design motion
> +spec.](https://material.io/guidelines/motion/movement.html)
> +The widget tree for the destination route is created in the `builder`
> +property.
> +
> +[Create a new Flutter example](/getting-started/) and
> +update it using the files from the
> +[GitHub directory.](https://github.com/flutter/website/tree/master/_includes/code/animation/flying_transition)
>
> The flying_transition folder didn't exist when I checked.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/website/pull/693#pullrequestreview-68500349>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGAHwpZMzz3KltgELZImSai0HYK8arYkks5srF2JgaJpZM4Py3xW>
> .
>

mravn-google commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+  position and size changes) from beginning to end,
+  moving the Hero along a curved path as described in the
+  Material motion spec. Flutter saves the size and location
+  of the Hero at the beginning and end of the transition.
```
position and size, then size and location

I think the main points are that 1) the destination route is created, 2) the framework detects that the destination route's widget tree contains a Hero with a matching tag, and then 3) the framework sets up tweens to interpolate between source position/size and destination position/size.

mravn-google commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+* Flutter enables the overlay, obscuring the current route.
+
+* Flutter creates the destination route, placing it at the
```
The destination Hero is part of the destination route's widget tree. I don't see how it can be available for tween calculations, if the destination route isn't created earlier than this.

HansMuller replied:
Right. At animation t=0.0 we create and layout the entire destination route "offscreen". This is done to figure out where the Hero is going to end up. Still at animation t=0.0 we move the destination Hero into the overlay and move the source hero offscreen. Then we start the hero animation. When the animation completes at t=1.0 we move both Heroes widgets back to the routes they came from (in other words at t=1.0 the destination hero moves from the overlay back to the destination route).

mravn-google commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+  but it's obscured by the overlay.
+
+* The destination route won't contain the destination Hero
+  until the transition is complete.
```
I don't get this.

HansMuller replied:
See above.

mravn-google commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+While the Hero is in-flight:
+
+* The same Animation object drives both
+  the route transition and the Hero animation.
```
What does this mean?

HansMuller replied:
This is a reminder that the route transition is animation that's happening at the same to the Hero animation is running.  The radial example uses the route's animation to fade in the destination route. 

Note also: this detail may be too much detail for most readers; we could leave it out.

InMatrix commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+   the same tag</strong>, typically an object that represents the
+   underlying data. For best results, the Heroes should have
+   virtually identical widget trees.
+1. (Optional) Place any Hero object that responds to user input,
```
I had the same question, though the Inkwell doesn't actually show the Material Design ink splash in the demo. Is it expected behavior?

InMatrix replied:
Or explain what alternative you can use for responding to tap, if you don't use Material Design.

InMatrix commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+(or using the device's back-to-previous-screen gesture)
+flies the flippers back to the original screen.
+
+<center><iframe width="560" height="315" src="https://www.youtube.com/embed/uF7_JfhBsww" frameborder="0" allowfullscreen></iframe></center>
```
Shams, can you add '&rel=0' to the end of the YouTube url to prevent distracting related videos from showing up after the demo video finishes? I found this trick here: https://wptavern.com/how-to-disable-related-videos-for-youtube-embeds-in-wordpress.

Sfshaza replied:
OMG! What a wonderful tip!!!

Sfshaza replied:
Oh bummer. Adding &rel=0 to the URL breaks the video.

InMatrix commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+  Heroes with matching tags.
+
+[Inkwell](https://docs.flutter.io/flutter/material/InkWell-class.html)
+: (Optional) Specifies what happens when tapping the Hero.
```
I was a little confused in what sense it's optional. It doesn't seem to be optional for this demo. Maybe in the general case, you can replace Inkwell with other widgets that respond to tap?

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+<img src="images/hero-transition-1.png" alt="Transition begins and Destination Hero appears in the overlay">
+
+Tapping the Hero's InkWell triggers the animation, and the
```
I guess I already mentioned this, I think it would be best to explain this section without mentioning the ink well.

I think it would be OK to say that the Hero animation is triggered by a Navigator push even though the reader will have to figure out that that happens because of a InkWell tap handler.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+* Flutter places the destination Hero in the overlay, at the
+  same location and size as the _source_ Hero.
+
+* Flutter enables the overlay, obscuring the current route.
```
The overlay doesn't obscure anything. Adding a Hero to the overlay just changes it's Z-order so that it appears on top of all routes.

Note that while the Hero animation is underway the destination Route is animating into view. That works because it no longer contains the Hero (since we moved its Hero to the overlay).

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+Before transition, the source Hero waits in the source route's widget
+tree. The destination route does not yet exist, and the overlay
+is blank and disabled.
```
The overlay is just empty.

InMatrix commented on _includes/code/animation/radial_transition/main.dart:
```diff
@@ -0,0 +1,180 @@
+
+void main() {
+  runApp(new MaterialApp(home: new RadialExpansionDemo()));
+}
```
When I ran this code, I couldn't load the three images. I copied the assets and the pubspec.yaml file. I was getting these errors:

```
[VERBOSE-2:dart_error.cc(16)] Unhandled exception:
	Unable to load asset: AssetManifest.json
	#0      PlatformAssetBundle.load (package:flutter/src/services/asset_bundle.dart:221:7)
	<asynchronous suspension>
	#1      CachingAssetBundle._fetchString (package:flutter/src/services/asset_bundle.dart:160:33)
	<asynchronous suspension>
	#2      CachingAssetBundle.loadString (package:flutter/src/services/asset_bundle.dart:156:12)
	#3      CachingAssetBundle.loadStructuredData (package:flutter/src/services/asset_bundle.dart:184:5)
	#4      AssetImage.obtainKey (package:flutter/src/services/image_resolution.dart:167:18)
	#5      ImageProvider.resolve (package:flutter/src/services/image_provider.dart:264:5)
	#6      _ImageState._resolveImage (package:flutter/src/widgets/image.dart:468:33)
	#7      _ImageState.didChangeDependencies (package:flutter/src/widgets/image.dart:449:5)
	#8      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3666:12)
	#9      ComponentElement.mount (package:flutter/src/widgets/framework.dart:3525:5)
	#10     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2890:14)
	#11     Element.updateChild (package:flutter/src/widgets/framework.dart:2693:12)
	#12     _LayoutBuilderElement._layout.<anonymous closure> (package:flutter/src/widgets/layout_builder.dart:119:18)
	#13     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2231:19)
	#14     _LayoutBuilderElement._layout (package:flutter/src/widgets/layout_builder.dart:107:11)
	#15     RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:2058:58)
	#16     PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:1148:15)
	#17     RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:2058:13)
	#18     _RenderLayoutBuilder.performLayout (package:flutter/src/widgets/layout_builder.dart:207:5)
	#19     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#20     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#21     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#22     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#23     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#24     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#25     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#26     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#27     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:1005:24)
	#28     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#29     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#30     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:1005:24)
	#31     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#32     RenderConstrainedBox.performLayout (package:flutter/src/rendering/proxy_box.dart:257:13)
	#33     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#34     RenderPositionedBox.performLayout (package:flutter/src/rendering/shifted_box.dart:373:13)
	#35     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#36     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#37     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:1005:24)
	#38     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#39     RenderConstrainedBox.performLayout (package:flutter/src/rendering/proxy_box.dart:257:13)
	#40     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#41     RenderFlex.performLayout (package:flutter/src/rendering/flex.dart:737:15)
	#42     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#43     RenderPositionedBox.performLayout (package:flutter/src/rendering/shifted_box.dart:373:13)
	#44     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#45     RenderPadding.performLayout (package:flutter/src/rendering/shifted_box.dart:197:11)
	#46     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#47     MultiChildLayoutDelegate.layoutChild (package:flutter/src/rendering/custom_layout.dart:124:11)
	#48     _ScaffoldLayout.performLayout (package:flutter/src/material/scaffold.dart:91:7)
	#49     MultiChildLayoutDelegate._callPerformLayout (package:flutter/src/rendering/custom_layout.dart:194:7)
	#50     RenderCustomMultiChildLayoutBox.performLayout (package:flutter/src/rendering/custom_layout.dart:338:14)
	#51     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#52     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#53     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#54     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#55     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:1005:24)
	#56     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#57     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#58     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#59     RenderStack.performLayout (package:flutter/src/rendering/stack.dart:466:15)
	#60     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#61     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#62     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#63     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#64     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#65     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#66     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#67     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#68     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#69     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#70     RenderOffstage.performLayout (package:flutter/src/rendering/proxy_box.dart:2749:26)
	#71     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#72     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#73     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#74     RenderStack.performLayout (package:flutter/src/rendering/stack.dart:466:15)
	#75     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#76     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#77     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#78     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#79     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#80     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#81     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#82     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#83     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#84     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107:13)
	#85     RenderObject.layout (package:flutter/src/rendering/object.dart:1962:7)
	#86     RenderView.performLayout (package:flutter/src/rendering/view.dart:142:13)
	#87     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1837:7)
	#88     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1126:18)
	#89     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:262:19)
	#90     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:581:22)
	#91     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:200:5)
	#92     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:713:15)
	#93     BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:649:9)
	#94     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/rendering/binding.dart:286:20)
	#95     Timer._createTimer.<anonymous closure> (dart:async-patch/dart:async/timer_patch.dart:21)
	#96     _Timer._runTimers (dart:isolate-patch/dart:isolate/timer_impl.dart:367)
	#97     _Timer._handleMessage (dart:isolate-patch/dart:isolate/timer_impl.dart:401)
	#98     _RawReceivePortImpl._handleMessage (dart:isolate-patch/dart:isolate/isolate_patch.dart:163)
```

Flutter Doctor output:
```
[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-US, channel master)
    • Flutter at /Users/taodong/flutter
    • Framework revision 17db2d2335 (2 hours ago), 2017-10-13 16:38:24 -0700
    • Engine revision 248af2d5b6
    • Tools Dart version 1.25.0-dev.11.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /Users/taodong/Library/Android/sdk
    • Platform android-25, build-tools 25.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    • ios-deploy 1.9.2
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.2.3)
    • Flutter plugin version 18.1
    • Dart plugin version 172.3757.55

[✓] Connected devices
    • iPhone 6s • B532439D-E9C6-4B65-A6C4-44388EFC0B8E • ios • iOS 10.3 (simulator)
```

InMatrix replied:
I tried the code on an Android physical device, and it was fine. So this has something to do with the iOS Simulator.

Sfshaza replied:
I have seen that in the Simulator, but I find if I try two or three times, it works fine. If you can reproduce it, can you file a bug against Simulator?

InMatrix replied:
This has been fixed in the latest master.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+* As the Hero flies, its rectangular bounds are animated using
+  [Tween&lt;Rect&gt;,](https://docs.flutter.io/flutter/animation/Tween-class.html)
+  accessible via the Hero's `createRectTween` property.
```
createRectTween isn't a RectTween it's a callback that creates a RectTween.

I think you could just say "specified with" instead of accessible via.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+---
+
+<img src="images/hero-transition-2.png" alt="The Hero flies in the overlay to its final position and size">
```
This diagram isn't really correct. During the Hero animation both routes are visible although neither one contains its Hero. While the Hero is flying the destination route is animating into place (using whatever route transition animation was defined for the destination route).

It might help to diagram a few widgets (like in light gray) on the source and destination routes.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+---
+
+<img src="images/hero-transition-1.png" alt="Transition begins and Destination Hero appears in the overlay">
```
It might be better if this diagram showed ghost rectangles that highlighted where the Heroes were on the source and destination routes.

InMatrix commented:
Should the formula at the center be `2 x maxRadius/sqrt(2)`?

![image](https://user-images.githubusercontent.com/348942/31571279-34553dce-b044-11e7-97b0-6b842ef939ed.png)

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+* Flutter moves the Hero widget from the overlay to the
+  destination route.
+
+* Flutter disables the (now blank) overlay.
```
I wouldn't say this. We don't disable overlays and the diagram makes it clear that it's empty.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+[Hero](https://docs.flutter.io/flutter/widgets/Hero-class.html)
+: The "shared element" that flies from the source to the destination
+  route. The Hero itself doesn't fly, it's actually the image that
```
This isn't right, it implies that we're flying a snapshot of the Hero or something.

What really happens is that we animate the Hero widget's bounds. That's really all there is to that.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+: The "shared element" that flies from the source to the destination
+  route. The Hero itself doesn't fly, it's actually the image that
+  represents the Hero, a child of the Hero.
+  Also, the Hero isn’t actually shared in the implementation.
```
If we stopped calling this topic "shared ...." this wouldn't need saying.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+and behavior when tapped. The PhotoHero builds the following
+widget tree:
+
+<img src="images/photohero-class.png" alt="widget tree for the PhotoHero class">
```
I think this diagram (or some text nearby) is the only place where it's worth mentioning the InkWell and the rationale for the Material widget (we don't want the splash to be left behind).

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+  flies to its destination.
+* Placing everything in a SizedBox and setting it's width
+  to the passed-in value, determines the size
+  of the image at the end of the animation.
```
Really it's the Hero's size at start and end of the animation. A PhotoHero is created for the source and Destination routes.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+  to the passed-in value, determines the size
+  of the image at the end of the animation.
+* Setting the Image's `fit` property to `BoxFit.contain`,
+  ensures that the image is as large as possible during the
```
as large as possible without changing its aspect ratio

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+* When the user taps the InkWell containing the source Hero,
+  the code creates a new MaterialPageRoute (in `onTap()`) and pushes
+  it to the Navigator’s stack, triggering the animation.
+* The anonymous builder for the new route creates a Container with the
```
Referring to this as anonymous isn't wrong but it doesn't convey very much. You're referring to the destination MaterialRoute's page `builder` function.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+  it to the Navigator’s stack, triggering the animation.
+* The anonymous builder for the new route creates a Container with the
+  destination PhotoHero. The Container's alignment determines where
+  the destination Hero lands.
```
That sounds like something I'd say :-).  The point here is that the Container is as big as the route (less its AppBar) and the Container's alignment - and its padding - defines where the PhotoHero is located within the route and below the AppBar.

It would probably be enough to say that the Container is used to position the PhotoHero in the route's top left corner, below the AppBar.

HansMuller commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+The following diagram shows the clipped image at the beginning
+(`t = 0.0`), and the end (`t = 1.0`) of the animation.
+
+<img src="images/radial-transition.png" alt="visual diagram of
```
You could achieve total symmetry maximization by reusing the ClipOval label in the center to point to the clip oval on the right :-)

Sfshaza replied:
LOL. Yes, I fixed it!

Sfshaza commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+---
+layout: page
+title: "Shared Element Transitions"
```
This was an early request, but I think you're right. It seems to add confusion.

Sfshaza commented on animations/shared-element-transitions/index.md:
```diff
@@ -0,0 +1,684 @@
+
+<aside class="alert alert-info" markdown="1">
+**Terminology:**
+A _Route_ describes a page or screen in a Flutter app. These terms
```
I hadn't seen this page! Is it new?

Sfshaza commented:
I *believe* I have incorporated all feedback. The page, renamed, is now staged here:

https://sz-flutter.firebaseapp.com/animations/hero-animations/

@HansMuller, InkWell is less discussed now, but should it be even less discussed? I've tweaked the 4-step animation diagrams. Do they seem ok?

Thanks, everyone, for your feedback!

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+</aside>
+
+You can create this animation in Flutter with Hero objects.
+When a new route appears, a hero animates from the current route to
```
I think it would be helpful to clarify that "appears" means that the rest of the route (less the Hero which is flying along its own path) animates into view. This is pretty plain to see in the video.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+
+The following video (recorded at slow speed),
+shows an example of a radial hero animation. At the start,
+a row of three images appears at the bottom of the route.
```
images => circular images

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+The following video (recorded at slow speed),
+shows an example of a radial hero animation. At the start,
+a row of three images appears at the bottom of the route.
+Tapping any of the circles flies that image to a new route
```
circles => circular images I guess. It seems important to use the same term to refer to the demo's heroes.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+
+<b> <a id="whats-the-point" class="anchor" href="#whats-the-point" aria-hidden="true"><span class="octicon octicon-link"></span></a>What's the point?</b>
+
+* Use two hero widgets with matching tags to implement the transition.
```
transition => animation, since the title is "... hero animation"

HansMuller replied:
two hero widgets in different routes but with matching tags

Saying all that might help the reader understand the connection to the following point about the Navigator and its stack of routes

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+* Use two hero widgets with matching tags to implement the transition.
+* The Navigator manages a stack containing the app’s routes.
+* Pushing a route on or popping a route from the Navigator’s stack
+  triggers the transition.
```
transition => animation per the earlier rationale

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+* The Navigator manages a stack containing the app’s routes.
+* Pushing a route on or popping a route from the Navigator’s stack
+  triggers the transition.
+* The Flutter framework calculates a rectangle tween that defines
```
Maybe link to https://docs.flutter.io/flutter/animation/RectTween-class.html since you're introducing a term here?

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+Hero animation code has the following structure:
+
+1. Define a starting Hero object, referred to as the _source
+   hero_. The hero specifies its location,
```
The hero doesn't specify its size or location, they're just a side effect of laying out the Hero widget.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+1. Create a route that contains the destination hero.
+   The destination route defines the widget tree that exists
+   at the end of the animation.
+1. Trigger the animation by pushing the route on the Navigator's stack.
```
the route => the destination route

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+   for each pair of heroes with matching tags in the source
+   and destination routes.
+
+Flutter calculates the tweens required to transition the
```
the tweens => the tween

You might clarify this part a little to emphasize that the tween is used to animate the bounds of the Hero

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+
+## Behind the scenes
+
+The following describes how Flutter performs the transition.
```
"transition" may not be clear to the reader, at least not yet. Might be helpful to explain that you're talking about the transition from one route to the other...

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+
+---
+
+<img src="images/hero-transition-1.png" alt="Transition begins and Destination hero appears in the overlay">
```
This isn't really right. We start the hero animation by putting the source hero offscreen.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+At t=0.0, Flutter does the following, mostly offscreen:
+
+* Calculates the transition (both position and size changes)
+  from beginning to end, moving the hero along a curved path as
```
We don't move the hero, there's no path involved here. We just layout the destination route - off screen - and then lookup where the hero ended up.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+* Moves the source hero offscreen.
+
+The framework kicks off the animation. When the animates
+completes (at t=1.0), Flutter moves both heroes back to the
```
The material about what happens when the flight completes is repeated below. I don't think it's needed here.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+
+[Hero](https://docs.flutter.io/flutter/widgets/Hero-class.html)
+: The widget that flies from the source to the destination route.
+  Define one Hero for the start of the transition and another for
```
Define one Hero for the source route and another for the destination route ...

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+
+* When the user taps the InkWell containing the source hero,
+  the code creates a new MaterialPageRoute (in `onTap()`) and pushes
+  it to the Navigator’s stack, triggering the animation.
```
It would be helpful to remind the reader that this route we've pushed is now the "destination".

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+* When the user taps the InkWell containing the source hero,
+  the code creates a new MaterialPageRoute (in `onTap()`) and pushes
+  it to the Navigator’s stack, triggering the animation.
+* The Container positions the PhotoHero in the route's top-left
```
route's => destination route's

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+  it to the Navigator’s stack, triggering the animation.
+* The Container positions the PhotoHero in the route's top-left
+  corner, below the AppBar.
+* The `onTap()` method for the destination PhotoHero pops the
```
destination => destination route's

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+<img src="images/radial-hero-animation.png" alt="visual diagram of
+radial transformation from beginning to end">
+
+The blue gradient (representing the image), indicates where these
```
these => the clip shapes

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+The following diagram shows the clipped image at the beginning
+(`t = 0.0`), and the end (`t = 1.0`) of the animation.
+
+<img src="images/radial-hero-animation.png" alt="visual diagram of
```
s = 2 * maxRadius / sqrt(2)

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,670 @@
+<ul markdown="1">
+<li markdown="1">The hero wraps the RadialExpansion widget.
+</li>
+<li markdown="1">As the hero flies, the RadialExpansion widget's
```
I think it would be helpful to explain that as the hero flies, its size changes, and because it constrains its child's size, the RadialExpansion widget changes size to match.

Sfshaza commented:
Thanks, @HansMuller! Feedback incorporated.

Staged here: https://sz-flutter.firebaseapp.com/animations/hero-animations/

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,666 @@
+in a Flutter app.
+</aside>
+
+You can create this animation in Flutter with Hero objects.
```
Hero widgets

I noticed that you've used "Hero object" pretty consistently. Not sure why you'd want to refer to them as just objects?

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,666 @@
+At t=0.0, Flutter does the following:
+
+* Lays out the destination route&mdash;off screen&mdash;using the
+  curved path as described in the Material motion spec.
```
The path doesn't have anything to do with laying out the destination route. Only the destination's route layout defines where the hero ends up.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,666 @@
+
+---
+
+<img src="images/hero-transition-2.png" alt="The hero flies in the overlay to its final position and size">
```
The diagram captions aren't quite right. The source route isn't necessarily obscured during the entirety of the hero's flight and the destination route is visible. You can see as much in your first demo.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,666 @@
+
+* The destination hero appears in its final position in
+  the destination route.
+
```
I think you have to admit that the source hero is restored to its route, since you explained that it's moved to the overlay at the flight's outset.

HansMuller commented on animations/hero-animations/index.md:
```diff
@@ -0,0 +1,666 @@
+  the code creates the destination route using MaterialPageRoute.
+  Pushing the destination route to the Navigator’s stack
+  triggers the animation.
+* The Container positions the PhotoHero in the destinatin route's
```
destinatin => destination
